File: kotlin/untitled/src/main/VendingMachine.java

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class VendingMachine {
    public static void main(String[] args) {
        List<Product> products = List.of(new Product(Type.COKE, 10),
                new Product(Type.CHIPS, 20));
        Vending vending = new Vending();

        addProducts(products, vending);
        vending.selectProduct(Type.COKE);

        List<Coin> coins = new ArrayList<>(List.of(Coin.TEN, Coin.TWENTY_FIVE));
        Map<Coin, Integer> coinIntegerMap = vending.insertCoin(coins);
        System.out.println("returned Coins :" + coinIntegerMap);

        Optional<Product> product = vending.dispenseProduct();

        product.ifPresent(System.out::println);
    }

    public static void addProducts(List<Product> products, Vending machine){
        for(Product p : products) {
            machine.addProduct(p);
        }
    }
}

File: kotlin/techverito_kotlin/src/main/VendingMachine.kt
import Coin.*
import java.util.*

class VendingMachine(private val products: ArrayList<Product> = ArrayList()) {
    private var selectedProduct: Product? = null
    private var coinSum: Int = 0
    private val coinInventory = mutableMapOf(ONE to 10, FIVE to 10, TEN to 10, TWENTY_FIVE to 10)

    fun addProducts(product: Product) {
        products.add(product)
    }

    fun selectProduct(type: Type) {
        selectedProduct = products.find { it.type == type }
        println("SelectedProduct is $selectedProduct")
    }

    fun dispenseProduct(): Product? {
        val dispenseProduct = products.find { it.type == selectedProduct?.type }
        println("Dispensing Product $dispenseProduct")
        return dispenseProduct
    }

    fun insertCoin(coins: List<Coin>): Map<Coin, Int> {
        coinSum = coins.map { it.num }.reduce { acc, next -> acc + next }
        println("coinSum $coinSum")

        if (coinSum < selectedProduct!!.price) {
            throw RuntimeException("Please enter more coins")
        }

        if (selectedProduct!!.price < coinSum) {
            return returnChange()
        }
        return Collections.emptyMap()
    }

    private fun returnChange(): Map<Coin, Int> {
        var excessAmount = coinSum - selectedProduct!!.price
        val returnCoins = HashMap<Coin, Int>()

        //check for highest denomination then give the coins
        if (coinInventory.getValue(TWENTY_FIVE) * 25 > coinSum) {
            val coins_25 = excessAmount / TWENTY_FIVE.num
            excessAmount = excessAmount % TWENTY_FIVE.num * coins_25
            coinInventory.computeIfPresent(TWENTY_FIVE) { key, value -> value - coins_25 }
            returnCoins.put(TWENTY_FIVE,coins_25)
        }
        return returnCoins
    }
}